pkg/storage/local/local.go

```diff
@@ -1477,12 +1491,131 @@ func ensureDir(dir string, log zerolog.Logger) error {
 }
 
 func (is *ImageStoreLocal) garbageCollect(dir string, repo string) error {
+	var lockLatency time.Time
+
 	oci, err := umoci.OpenLayout(dir)
 	if err := test.Error(err); err != nil {
 		return err
 	}
 	defer oci.Close()
 
+	// gc untagged manifests and signatures
+	index, err := oci.GetIndex(context.Background())
+	if err != nil {
+		return err
+	}
+
+	referencedByImageIndex := []string{}
+
+	for _, desc := range index.Manifests {
+		if desc.MediaType == ispec.MediaTypeImageIndex {
+			is.Unlock(&lockLatency)
+			indexImage, err := storage.GetImageIndex(is, repo, desc.Digest, is.log)
+			is.Lock(&lockLatency)
+			if err != nil {
+				return err
+			}
+
+			for _, indexDesc := range indexImage.Manifests {
+				referencedByImageIndex = append(referencedByImageIndex, indexDesc.Digest.String())
+			}
+		}
+	}
+
+	cosignDescriptors := []ispec.Descriptor{}
+	notationDescriptors := []ispec.Descriptor{}
+
+	for _, desc := range index.Manifests {
+		// skip manifests referenced in image indexex
+		if common.Contains(referencedByImageIndex, desc.Digest.String()) {
+			continue
+		}
+
+		// remove untagged images
+		if desc.MediaType == ispec.MediaTypeImageManifest {
+			tag, ok := desc.Annotations[ispec.AnnotationRefName]
+			if !ok {
+				// remove manifest if it's older than gc.delay
+				canGC, err := isBlobOlderThan(is, repo, desc.Digest, is.gcDelay)
+				if err != nil {
+					return err
+				}
+
+				if canGC {
+					storage.RemoveManifestDescByReference(&index, desc.Digest.String())
+
+					err := oci.PutIndex(context.Background(), index)
+					if err != nil {
+						return err
+					}
+				}
+			}
+
+			// gather cosign signatures
+			if strings.HasPrefix(tag, "sha256-") && strings.HasSuffix(tag, remote.SignatureTagSuffix) {
+				cosignDescriptors = append(cosignDescriptors, desc)
+			}
+		}
+
+		// gather notation signatures
+		if desc.MediaType == artifactspec.MediaTypeArtifactManifest {
+			notationDescriptors = append(notationDescriptors, desc)
+		}
+	}
+
+	foundSubject := false
+
+	for _, cosignDesc := range cosignDescriptors {
+		// check if we can find the manifest which the signature points to
+		for _, desc := range index.Manifests {
+			subject := fmt.Sprintf("sha256-%s.%s", desc.Digest.Encoded(), remote.SignatureTagSuffix)
+			if subject == cosignDesc.Annotations[ispec.AnnotationRefName] {
+				foundSubject = true
+			}
+		}
+
+		if !foundSubject {
+			// remove manifest
+			storage.RemoveManifestDescByReference(&index, cosignDesc.Digest.String())
+
+			err := oci.PutIndex(context.Background(), index)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	foundSubject = false
+
+	for _, notationDesc := range notationDescriptors {
+		// check if we can find the manifest which the signature points to
+		artManifest, err := getArtifactManifestByDesc(dir, notationDesc, is.log)
+		if err != nil {
+			return err
+		}
+
+		if artManifest.ArtifactType != "application/vnd.cncf.notary.v2.signature" {
```

Comment:
Isn't there a constant defined in the notation library for this?

Reply:
fixed, thanks